lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU

```diff
@@ -129,9 +139,17 @@ ffNoGoal(
     This action sets bForwardEnable and bBackwardEnable based on
     the current position and the calculated bounds.
 *)
-IF bValidGoal AND bEnable THEN
-    bForwardEnabled := stMotionStage.stAxisStatus.fActPosition < fUpperPos;
-    bBackwardEnabled := stMotionStage.stAxisStatus.fActPosition > fLowerPos;
+IF bLockBounds THEN
+    // Prevent forward/backward motion if the position of the axis is outside of the upper/lower bounds respectively.
+    // Prevent forward/backward motion if the command to the axis is outside of the upper/lower bounds respectively.
+    bForwardEnabled := stMotionStage.stAxisStatus.fActPosition < fUpperPos AND stMotionStage.fPosition < fUpperPos;
+    bBackwardEnabled := stMotionStage.stAxisStatus.fActPosition > fLowerPos AND stMotionStage.fPosition > fLowerPos;
+
+    IF stMotionStage.nErrorId = 16#4223 AND (NOT bForwardEnabled OR NOT bBackwardEnabled) THEN
+        // Cannot move forward or backward, overwrite the error with a custom error message to give more context.
+        stMotionStage.sCustomErrorMessage := CONCAT(CONCAT(CONCAT(
+            'Limits exceeded for most recent PMPS position state: ', LREAL_TO_FMTSTR(fLowerPos,5,TRUE)), ' < pos < '), LREAL_TO_FMTSTR(fUpperPos,5,TRUE));
```

Comment:
It's a somewhat absurd edge case, but this error isn't necessarily correct if the user wants to move positive and the PMPS restriction is "bForwardEnable = True, bBackwardEnabled = False" but instead they are hitting the positive limit to bring forth the 16#4223 error, or otherwise the general enable is False.

I'd suggest matching the directional enables with the last setpoint vs the current position.

Reply:
I made it so that this special error message will only display if the bAllForwardEnable/bAllBackwardEnable from stMotionStage are true and they line up with the correct PMPS direction enable

Reply:
Looks pretty comprehensive to me :+1: